cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/replication/BasicPullStrategy.java

```diff
@@ -204,36 +235,48 @@ private void replicate()
             );
             logger.info(msg);
             long batchStartTime = System.currentTimeMillis();
-
-            ChangesResultWrapper changeFeeds = this.nextBatch();
-            int batchChangesProcessed = 0;
-
-            // So we can check whether all changes were processed during
-            // a log analysis.
-            msg = String.format(
-                    "Batch %s contains %s changes",
-                    this.batchCounter,
-                    changeFeeds.size()
-            );
-            logger.info(msg);
-
-            if (changeFeeds.size() > 0) {
-                batchChangesProcessed = processOneChangesBatch(changeFeeds);
-                documentCounter += batchChangesProcessed;
+            BatchProcessor batchProcessor = new BatchProcessor();
+
+            // Call the BatchProcessor to fetch the next batch from the changes feed and then process
+            // them. Retry this in case the changes feed is inconsistent with the revisions available.
+            // See BasicPullStrategyTest#pull_changesNewerThanOpenRevs for more explanation of this
+            // scenario.
+            //
+            // Essentially the BatchProcessor is just a wrapper for the Callable#call method and so
+            // we can access changeFeeds and batchChangesProcessed fields after successful execution
+            //
+            // Note that it is OK to call nextBatch() repeatedly for each retry without missing any
+            // batches - this is because we know that the checkpoint has not been updated.
+            //
+            // Since RetriableTask#call can throw Exception, we have to re-throw the exceptions we
+            // are expecting, plus wrapping and throwing any that we aren't.
+            try {
+                new RetriableTask<BatchProcessor>(batchProcessor).call();
+            } catch (ExecutionException e) {
```

Comment:
I agree. I'm at a loss for a better strategy - my guess is that anything using `Callable` or similar has to do the same because the `call()` method is capable of throwing any `Exception`.